fix xcom push not working in `SparkKubernetesOperator`

kesem0811 opened the pull request:
Fixes: https://github.com/apache/airflow/issues/39184 where SparkKubernetesOperator tasks hang indefinitely because no sidecar is injected to read /airflow/xcom/return.json.

kesem0811 committed:
add xcom sidecar

kesem0811 committed:
this version is working

kesem0811 committed:
final version without test changes

kesem0811 committed:
delete prints

kesem0811 committed:
your commit message

kesem0811 commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/operators/spark_kubernetes.py:
```diff
@@ -293,10 +293,24 @@ def client(self) -> CoreV1Api:
+                )
+                self.template_body["spark"]["spec"]= driver_with_xcom_template
+            except KeyError as e:
+                raise AirflowException("Spec missing in SparkApplication template") from e
```
Why?
In all the file, when something doesn't work as expected, an AirflowException is raised

eladkal replied:
lets switch to KeyError

kesem0811 committed:
cr fix

kesem0811 committed:
cr fix

kesem0811 committed:
your message

kesem0811 committed:
Merge branch 'main' of https://github.com/kesem0811/airflow into feature/fix_xcom_push_in_SparkKubernetesOperator

kesem0811 committed:
resolve conflicts

kesem0811 committed:
run ruff format

kesem0811 committed:
Merge upstream main into feature/fix_xcom_push_in_SparkKubernetesOperator

- Resolved merge conflict in spark_kubernetes.py
- Preserved xcom sidecar functionality from local changes
- Integrated upstream improvements including deferrable support and better reattach logic
- Combined both approaches: upstream refactoring + local xcom features

kesem0811 committed:
fix: resolve mypy type error in xcom_sidecar.py

- Add proper type checking for dict vs V1Pod inputs in add_xcom_sidecar()
- Convert dict inputs to V1Pod objects to ensure .spec attribute access
- Maintains backward compatibility while fixing type safety
- Resolves: Item 'dict[Any, Any]' has no attribute 'spec' [union-attr]

kesem0811 commented:
the static checks failed because 
The image /mnt/ci-image-save-v3-linux_amd64-3.10.tar does not exist.
I love to make the tests run again? 
@romsharon98 @eladkal

romsharon98 committed:
Merge branch 'main' into feature/fix_xcom_push_in_SparkKubernetesOperator

kesem0811 committed:
style: apply ruff formatting to spark_kubernetes.py

kesem0811 committed:
ruff formating

kesem0811 committed:
Merge branch 'main' into feature/fix_xcom_push_in_SparkKubernetesOperator

kunaljubce commented:
@kesem0811 Can you also fix the static checks CI error? Should be autofixed by your [prek](https://github.com/apache/airflow/blob/main/contributing-docs/03_contributors_quick_start.rst#configuring-prek) pre-commit hooks, assuming you have them installed.

Lee-W commented:
will also need your help resolving the conflict. Thanks!

potiuk commented:
@kesem0811 This draft PR has had no activity for 3 weeks. Closing to keep the queue clean.

You are welcome to reopen and continue when you're ready. If you'd like to pick it back up, please rebase onto the current `main` branch first.

---

_Note: This comment was drafted by an AI-assisted triage tool and may contain mistakes. Once you have addressed the points above, an Apache Airflow maintainer — a real person — will take the next look at your PR. We use this [two-stage triage process](https://github.com/apache/airflow/blob/main/contributing-docs/25_maintainer_pr_triage.md#why-the-first-pass-is-automated) so that our maintainers' limited time is spent where it matters most: the conversation with you._

kesem0811 commented:
Hi everyone - sorry  for not responding earlier. That’s on me, and I understand why the PR was closed.

I’ve opened a new PR that addresses the review feedback and completes the fix: https://github.com/apache/airflow/pull/68788
